### PR TITLE
[usage] add usage API for credit notes

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
As part of providing a means in the admin dashboard to issue credits to cost centers.

## Related Issue(s)

Fixes #14594

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/usage/pkg/db/cost_center.go [hunk 1]
```diff
@@ -155,70 +155,39 @@ func (c *CostCenterManager) UpdateCostCenter(ctx context.Context, newCC CostCent
...
-			newCC.BillingCycleStart = common_db.NewVarCharTime(now)
-			// set an informative nextBillingTime, even though we don't manage Stripe billing cycle
-			newCC.NextBillingTime = common_db.NewVarCharTime(now.AddDate(0, 1, 0))
-		}
 	} else if isTeam {
-		// Billing strategy is Other, and it remains unchanged
-		if existingCC.BillingStrategy == CostCenter_Other && newCC.BillingStrategy == CostCenter_Other {
-			// It is impossible for a team without Stripe billing to change their spending limit
-			if newCC.SpendingLimit != c.cfg.ForTeams {
-				return CostCenter{}, status.Errorf(codes.FailedPrecondition, "teams without a subscription cannot change their spending limit")
```
Comment (svenefftinge) [inline components/usage/pkg/db/cost_center.go:197]: This invariant is no longer true, as admins are able to adjust the spending limit for 'free' accounts going forward.

File: components/usage/pkg/apiv1/usage.go [hunk 2]
```diff
@@ -453,3 +454,48 @@ func NewUsageService(conn *gorm.DB, pricer *WorkspacePricer, costCenterManager *
...
+	userId, err := uuid.Parse(req.UserId)
+	if err != nil {
+		return nil, fmt.Errorf("The user id is not a valid UUID. %w", err)
+	}
+
+	usage := db.Usage{
+		ID:            uuid.New(),
+		AttributionID: attributionId,
+		Description:   note,
+		CreditCents:   db.NewCreditCents(float64(req.Credits * -1)),
```
Comment (geropl) [inline components/usage/pkg/apiv1/usage.go:485]: :cloud: Would it make sense to check for the sign of `req.Credits`? Or is this call supposed to support both?

File: components/usage/pkg/db/cost_center.go [hunk 3]
```diff
@@ -155,70 +155,39 @@ func (c *CostCenterManager) UpdateCostCenter(ctx context.Context, newCC CostCent
...
+	}
 
-			newCC.BillingCycleStart = common_db.NewVarCharTime(now)
-			// set an informative nextBillingTime, even though we don't manage Stripe billing cycle
-			newCC.NextBillingTime = common_db.NewVarCharTime(now.AddDate(0, 1, 0))
+	// Upgrading to Stripe
+	if existingCC.BillingStrategy == CostCenter_Other && newCC.BillingStrategy == CostCenter_Stripe {
+		err := c.BalanceOutUsage(ctx, attributionID)
+		if err != nil {
+			return CostCenter{}, err
```
Comment (geropl) [inline components/usage/pkg/db/cost_center.go:185]: `nil, err`?

File: components/usage/pkg/apiv1/usage.go (hunk 2 above)
Comment (svenefftinge) [reply components/usage/pkg/apiv1/usage.go:485]: it supports both

File: components/usage/pkg/db/cost_center.go (hunk 1 above)
Comment (geropl) [reply components/usage/pkg/db/cost_center.go:197]: This check (similarly the analog for users) were meant to block users to (inadvertently) cheat. How is that meant to be enforced now? :thinking:

File: components/usage/pkg/db/cost_center.go (hunk 3 above)
Comment (svenefftinge) [reply components/usage/pkg/db/cost_center.go:185]: I will change in [the follow up PR](https://github.com/gitpod-io/gitpod/pull/14594), if nothing else comes up.

File: components/usage/pkg/db/cost_center.go (hunk 1 above)
Comment (svenefftinge) [reply components/usage/pkg/db/cost_center.go:197]: It's done here: https://github.com/gitpod-io/gitpod/blob/062ea9b07a951949ab4db986aadbad9af2c3127e/components/server/ee/src/workspace/gitpod-server-impl.ts#L2318-L2323

Comment (geropl) [reply components/usage/pkg/db/cost_center.go:197]: Ah thx. Need a workspace for better overview I guess :wink:

File: components/usage/pkg/db/usage.go [hunk 4]
```diff
@@ -82,6 +93,10 @@ type WorkspaceInstanceUsageData struct {
 	UserAvatarURL  string        `json:"userAvatarURL"`
 }
 
+type CreditNoteMetaData struct {
+	UserId string `json:userId`
```
Comment (geropl) [inline components/usage/pkg/db/usage.go:97]:
```suggestion
	UserID string `json:"userId"`
```

**Review by geropl — APPROVED** [review]
Code LGTM, thx! :pray: 

I got two nits [[1](https://github.com/gitpod-io/gitpod/pull/14769#discussion_r1026200114), [2](https://github.com/gitpod-io/gitpod/pull/14769#discussion_r1026146135)], and a more general concern [here](https://github.com/gitpod-io/gitpod/pull/14769#discussion_r1026205474). Both neither of those should block this PR.

/hold in case you want to comment/adress here

**Pull request merged**